community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java

```diff
@@ -0,0 +1,316 @@
+                }
+            }
+            catch ( LegacyIndexNotFoundKernelException | EntityNotFoundException | PropertyNotFoundException e )
+            {
+                throw new AutoIndexingKernelException( e );
+            }
+            catch ( TokenNotFoundException e )
+            {
+                // TODO: TokenNotFoundException was added before there was a kernel. It should be converted to a
+                // KernelException now
+                throw new AutoIndexingKernelException( new PropertyKeyIdNotFoundKernelException( oldProperty
+                        .propertyKeyId(), e ) );
+            }
+        }
+    }
+
+    @Override
+    public void propertyRemoved( DataWriteOperations ops, long entityId, int propertyKey )
+            throws AutoIndexingKernelException
+    {
+        if ( enabled )
+        {
+            try
+            {
+                String name = propertyKeyLookup.getTokenById( propertyKey ).name();
+                if ( propertyKeysToInclude.get().contains( name ) )
+                {
+                    ensureIndexExists( ops );
+                    type.remove( ops, entityId );
+                }
+            }
+            catch ( LegacyIndexNotFoundKernelException | EntityNotFoundException e )
+            {
+                throw new AutoIndexingKernelException( e );
+            }
+            catch ( TokenNotFoundException e )
+            {
+                // TODO: TokenNotFoundException was added before there was a kernel. It should be converted to a
+                // KernelException now
+                throw new AutoIndexingKernelException( new PropertyKeyIdNotFoundKernelException( propertyKey, e ) );
+            }
+        }
+    }
+
+    @Override
+    public void entityRemoved( DataWriteOperations ops, long entityId ) throws AutoIndexingKernelException
+    {
+        if ( enabled )
+        {
+            try
+            {
+                ensureIndexExists( ops );
+                type.remove( ops, entityId );
+            }
+            catch ( LegacyIndexNotFoundKernelException | EntityNotFoundException e )
+            {
+                throw new AutoIndexingKernelException( e );
+            }
+        }
+    }
+
+    // Trap door needed to keep this as an enum
+    void replacePropertyKeysToInclude( List<String> propertyKeysToIncludeNow )
+    {
+        Set<String> copiedPropertyKeysToIncludeNow = new HashSet<>( propertyKeysToIncludeNow.size() );
+        copiedPropertyKeysToIncludeNow.addAll( propertyKeysToIncludeNow );
+        this.propertyKeysToInclude.set( copiedPropertyKeysToIncludeNow );
+    }
+
+    @Override
+    public void enabled( boolean enabled )
+    {
+        this.enabled = enabled;
+    }
+
+    @Override
+    public boolean enabled()
+    {
+        return enabled;
+    }
+
+    @Override
+    public void startAutoIndexingProperty( String propName )
+    {
+        propertyKeysToInclude.getAndUpdate( current -> {
```

Comment:
as soon as set will be released from atomicReference, this can be simply achieved by any concurrent set implementations if multi threaded access is required.